### ISSUE TITLE
Confusion between Big5 and Big5-HKSCS encodings

### ISSUE THREAD

**Issue by bhaible** [whatwg/encoding#75]
The current draft maps the labels "big5" and "big5-hkscs" to a single encoding, and the mapping table that it uses (index-big5.txt) does not match either of the widely used mapping tables for BIG5 and BIG5-HKSCS.

In detail:

Glibc and other software consider BIG5 and BIG5-HKSCS to be different.
BIG5 is mainly used in Taiwan, and evolves over time without formal standards. But Microsoft is a strong player in this area, as it provides the Windows CP950 encoding.
BIG5-HKSCS is mainly used in Hong Kong, and evolves through regular updates of the standard. The current version BIG5-HKSCS:2008 is found in http://www.ogcio.gov.hk/en/business/tech_promotion/ccli/terms/doc/e_hkscs_2008.pdf .

The mapping in index-big5.txt differs from the common BIG5 mapping (Microsoft CP950): It adds 5085 characters, at the positions 0x8862, 0x8864, 0x88A3, 0x88A5, 0x8740..0xA0FE, 0xA3C0..0xA3E0, 0xC6A1..0xC8FE, 0xFA40..0xFEFE.

The mapping in index-big5.txt differs from the one in BIG5-HKSCS:2008:
It has additional mappings for 0x8E69, 0x8E6F, 0x8E7E, 0x8EAB, 0x8EB4, 0x8ECD, 0x8ED0, 0x8F57, 0x8F69, 0x8F6E, 0x8FCB..0x8FCC, 0x8FFE, 0x906D, 0x907A, 0x90DC, 0x90F1, 0x91BF, 0x9244, 0x92AF..0x92B2, 0x92C8, 0x92D1, 0x9447, 0x94CA, 0x95D9, 0x9644, 0x96ED, 0x96FC, 0x9B76, 0x9B78, 0x9B7B, 0x9BC6, 0x9BDE, 0x9BEC, 0x9BF6, 0x9C42, 0x9C53, 0x9C62, 0x9C68, 0x9C6B, 0x9C77, 0x9CBC, 0x9CBD, 0x9CD0, 0x9D57, 0x9D5A, 0x9DC4, 0x9EA9, 0x9EEF, 0x9EFD, 0x9F60, 0x9F66, 0x9FCB, 0x9FD8, 0xA063, 0xA077, 0xA0D5, 0xA0DF, 0xA0E4, 0xA15A, 0xA1C3, 0xA1C5, 0xA1FE, 0xA240, 0xA2CC, 0xA2CE, 0xA3C0..0xA3E1, 0xC6CF, 0xC6D3, 0xC6D5, 0xC6D7, 0xC6DE..0xC6DF, 0xFA5F, 0xFA66, 0xFABD, 0xFAC5, 0xFAD5, 0xFB48, 0xFBB8, 0xFBF3, 0xFBF9, 0xFC4F, 0xFC6C, 0xFCB9, 0xFCE2, 0xFCF1, 0xFDB7, 0xFDB8, 0xFDBB, 0xFDF1, 0xFE52, 0xFE6F, 0xFEAA, 0xFEDD,
and differs in the mappings of 0xA145, 0xA14E, 0xA1C2, 0xA1E3, 0xA1F2..0xA1F3, 0xA241..0xA242, 0xA244, 0xA246..0xA247.

For details about the mapping tables, see
http://haible.de/bruno/charsets/conversion-tables/index.html
http://haible.de/bruno/charsets/conversion-tables/Big5.html
http://haible.de/bruno/charsets/conversion-tables/BIG5-HKSCS.html

In summary, it looks like this encoding is a clever merge between CP950 and BIG5-HKSCS, giving priority to the CP950 mapping value in those 11 positions where there is a conflict. This is very good for converting BIG5/BIG5-HKSCS texts of either kind to Unicode (the "decoder" part). However, when the "encoder" is used, it will likely generate HKSCS code points, which cannot be displayed in CP950. What shall the Windows users with this CP950 encoding do?

**Comment by annevk** [comment]
The encoder avoids emitting HKSCS code points through https://encoding.spec.whatwg.org/#index-big5-pointer, no?

**Comment by bhaible** [comment]
Thanks. I had not seen this detail.

But, still, the Big5 encoder may produce byte pairs in the ranges 0xA3C0..0xA3E0, 0xC6A1..0xC8FE, 0xFA40..0xFEFE. Between 1163 and 1183 characters, if I count correctly. These characters are not contained in Windows CP950. What shall the Windows users with this CP950 encoding do?

**Comment by hsivonen** [comment]
cc @foolip 

Some thoughts:

1. Taking the previous comment at face value, it does seem inconsistent to avoid generating some HKSCS byte sequences while not avoiding others.
2. An implementation compliant with this spec shipped in Firefox 43 on 2015-12-15. We haven't received bug report about the encoder behavior. (In contrast, before shipping the merged Big5/Big5-HKSCS decode, we tried simply decoding content labeled as Big5-HKSCS using the old Big5 decoder, and we got a bug report about that.)
3. It seems that there are some code points (I haven't worked out which, yet) that the Encoding Standard has Big5 encoder mappings for but glibc does not have Big5-HKSCS encoder mappings for. (Not necessarily cause for action on the spec side--just an observation.)

**Comment by hsivonen** [comment]
> What shall the Windows users with this CP950 encoding do?

I think that's not really the most relevant question, since Kernel32.dll probably isn't the most popular Web server encoding conversion back end.

The relevant question is whether the current encoder behavior breaks some server-side programs that expect form submissions in some notion of "Big5".

**Comment by hsivonen** [comment]
> It seems that there are some code points (I haven't worked out which, yet) that the Encoding Standard has Big5 encoder mappings for but glibc does not have Big5-HKSCS encoder mappings for. 

Oops. I misread what I saw. This was about decode after all.

glibc's iconv's big5-hkscs fails when decoding 0xC6 0xD3, which Encoding Standard decodes to U+65E0.

**Comment by foolip** [comment]
Here's my last attempt to summarize the state of things:
https://bugzilla.mozilla.org/show_bug.cgi?id=912470#c6
https://bugzilla.mozilla.org/show_bug.cgi?id=912470#c48

That broken link to lists.whatwg.org is https://lists.w3.org/Archives/Public/public-whatwg-archive/2012Apr/0082.html

The nature of the problem in this issue isn't entirely clear to me, is it encoder or decoder after all? @bhaible, can you elaborate on what should change?

> glibc's iconv's big5-hkscs fails when decoding 0xC6 0xD3, which Encoding Standard decodes to U+65E0.

This came up in https://lists.w3.org/Archives/Public/public-whatwg-archive/2012Apr/thread.html#msg95 too. Is the mapping currently in the spec a problem?

**Comment by hsivonen** [comment]
> > glibc's iconv's big5-hkscs fails when decoding 0xC6 0xD3, which Encoding Standard decodes to U+65E0.
>
> This came up in https://lists.w3.org/Archives/Public/public-whatwg-archive/2012Apr/thread.html#msg95 too. Is the mapping currently in the spec a problem?

I'm not aware of this being a practical problem on the Web. I just happened to notice this when I tried to benchmark encoding_rs vs. glibc.

On the high level, I'm reluctant to change anything regarding Big5 in the spec without evidence of breakage exhibited on the Web. (That is, abstract inconsistency with a particular converter library isn't in itself a reason to change the spec.)

**Comment by foolip** [comment]
That is my inclination as well. To change any specific code points, some analysis of its effect in the wild would be good, or reasons to think that the change doesn't matter in the wild.

**Comment by annevk** [comment]
@bhaible any final thoughts? I'm inclined to close this.

**Comment by annevk** [comment]
Closing this due to lack of sufficient new information to change opinions around.
